src/Passes/Source/Passes/ExpandStaticAllocation/ExpandStaticAllocation.hpp

```diff
@@ -2,9 +2,9 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-#include "Llvm.hpp"
+#include "Passes/QubitAllocationAnalysis/QubitAllocationAnalysis.hpp"
 
-#include "QubitAllocationAnalysis/QubitAllocationAnalysis.hpp"
+#include "Llvm/Llvm.hpp"
```

Comment:
Yes, order sort-of-matters. When you include in C++ it is the same as just copy/pasting the function into the place where you do the include. This means that if I have a header file `foo.hpp` that depends on iostream and a main file that includes the header and iostream:
```c++
#include<iostream>

#include "foo.h"
int main() { /* ... */ }
```
but I forgot to include `iostream` in `foo.hpp` itself, then everything will compile just fine, but `foo.hpp` will not compile on it's own. Hence if I make another program where I include `foo.hpp` and `iostream` afterwards
```c++
#include "foo.h"
#include<iostream>

int main() { /* ... */ }
```
everything will break. Therefore, the style guide (clang-format) is setup to order includes according to the latter as this will immediately reveal if you have forgotten to include a dependency in `foo.hpp`. As such, you could do without this requirement, but as you scale up your code base these types of missing headers will end up causing you a great deal of headaches.

I am not sure whether it is worthwhile to put a comment about the ordering in this specific file as clang format will automatically rearrange everything according to its liking. And probably, for this particular instance, it would not matter at all - it is just clang-format that insists.

Reply:
Makes sense. Since it was clang-format that rearranged them I'm less worried.

As an aside, I've seen corner cases where ordering matters in C++ includes such that changing the order of headers doesn't cause compilation to fail but does change behavior in a way that introduces a subtle bug. That's actually a technique that is used to frightening effect in some [obfuscated C contests](https://en.wikipedia.org/wiki/International_Obfuscated_C_Code_Contest).